File: app/src/test/java/tech/pauly/findapet/data/AnimalRepositoryTest.java
package tech.pauly.findapet.data;

import android.support.annotation.NonNull;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.Collections;

import io.reactivex.Single;
import io.reactivex.observers.TestObserver;
import tech.pauly.findapet.data.models.Age;
import tech.pauly.findapet.data.models.Animal;
import tech.pauly.findapet.data.models.AnimalListResponse;
import tech.pauly.findapet.data.models.AnimalSize;
import tech.pauly.findapet.data.models.AnimalType;
import tech.pauly.findapet.data.models.FetchAnimalsRequest;
import tech.pauly.findapet.data.models.Filter;
import tech.pauly.findapet.data.models.Sex;
import tech.pauly.findapet.data.models.StatusCode;

import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class AnimalRepositoryTest {

    @Mock
    private AnimalService animalService;

    @Mock
    private ObservableHelper observableHelper;

    private AnimalRepository subject;
    private AnimalListResponse animalListResponse;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        animalListResponse = mock(AnimalListResponse.class);
        when(animalService.fetchAnimals(anyString(),
                                        anyString(),
                                        anyString(),
                                        anyInt(),
                                        anyInt(),
                                        anyString(),
                                        anyString(),
                                        anyString(),
                                        anyString())).thenReturn(Single.just(animalListResponse));
        when(observableHelper.applySingleSchedulers()).thenReturn(single -> single);

        subject = new AnimalRepository(animalService, observableHelper);
    }

    @Test
    public void fetchAnimals_returnAnimalListNotNullOrEmpty_returnsAnimalListForCorrectAnimalTypeWithSchedulers() {
        FetchAnimalsRequest request = setupFilterRequest();
        when(animalListResponse.getAnimalList()).thenReturn(Collections.singletonList(mock(Animal.class)));

        TestObserver<AnimalListResponse> observer = subject.fetchAnimals(request).test();

        verify(animalService).fetchAnimals(eq("zipcode"),
                                           anyString(),
                                           eq("cat"),
                                           eq(0),
                                           eq(20),
                                           eq("M"),
                                           eq("Adult"),
                                           eq("L"),
                                           eq("Calico"));
        observer.assertValues(animalListResponse)
                .assertComplete();
        verify(observableHelper).applySingleSchedulers();
    }

    @Test
    public void fetchAnimals_returnAnimalListNull_returnExceptionForNoAnimals() {
        FetchAnimalsRequest request = setupFilterRequest();
        when(animalListResponse.getAnimalList()).thenReturn(null);

        TestObserver<AnimalListResponse> observer = subject.fetchAnimals(request).test();

        observer.assertError(new PetfinderException(StatusCode.ERR_NO_ANIMALS));
    }

    @Test
    public void fetchAnimals_returnAnimalListEmpty_returnExceptionForNoAnimals() {
        FetchAnimalsRequest request = setupFilterRequest();
        when(animalListResponse.getAnimalList()).thenReturn(Collections.emptyList());

        TestObserver<AnimalListResponse> observer = subject.fetchAnimals(request).test();

        observer.assertError(new PetfinderException(StatusCode.ERR_NO_ANIMALS));
    }

    @NonNull
    private FetchAnimalsRequest setupFilterRequest() {
        Filter filter = new Filter();
        filter.setSex(Sex.MALE);
        filter.setAge(Age.ADULT);
        filter.setSize(AnimalSize.LARGE);
        filter.setBreed("Calico");
        return new FetchAnimalsRequest(AnimalType.CAT, 0, "zipcode", filter);
    }
}

File: app/src/test/java/tech/pauly/findapet/data/AnimalRepositoryTest.kt
package tech.pauly.findapet.data

import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.mockk
import io.mockk.verify
import io.reactivex.Single
import io.reactivex.SingleTransformer
import org.junit.Before
import org.junit.Test
import tech.pauly.findapet.data.models.*

class AnimalRepositoryTest {

    @MockK
    private lateinit var animalService: AnimalService

    @MockK
    private lateinit var observableHelper: ObservableHelper

    @MockK
    private lateinit var animalListResponse: AnimalListResponse

    @MockK
    private lateinit var singleAnimalResponse: SingleAnimalResponse

    @MockK
    private lateinit var header: Header

    private lateinit var subject: AnimalRepository

    @Before
    fun setup() {
        MockKAnnotations.init(this)
        every { animalService.fetchAnimals(any(), any(), any(), any(), any(), any(), any(), any(), any()) } returns Single.just(animalListResponse)
        every { observableHelper.applySingleSchedulers<Any>() } returns SingleTransformer { it }
        every { animalListResponse.header } returns header
        every { singleAnimalResponse.header } returns header

        subject = AnimalRepository(animalService, observableHelper)
    }

    @Test
    fun fetchAnimals_returnAnimalListNotNullOrEmpty_returnsAnimalListForCorrectAnimalTypeWithSchedulers() {
        val request = setupFilterRequest()
        val status = mockk<Status> {
            every { code } returns StatusCode.PFAPI_OK
        }
        every { header.status } returns status
        every { animalListResponse.animalList } returns listOf(mockk())

        val observer = subject.fetchAnimals(request).test()

        observer.assertValues(animalListResponse).assertComplete()
        verify {
            animalService.fetchAnimals("zipcode",
                    any(),
                    "cat",
                    0,
                    20,
                    "M",
                    "Adult",
                    "L",
                    "Calico")
            observableHelper.applySingleSchedulers<Any>()
        }
    }

    @Test
    fun fetchAnimals_statusCodeOK_returnsOnNext() {
        val request = setupFilterRequest()
        every { animalListResponse.animalList } returns listOf(mockk())
        val status = mockk<Status> {
            every { code } returns StatusCode.PFAPI_OK
        }
        every { header.status } returns status

        val observer = subject.fetchAnimals(request).test()

        observer.assertValues(animalListResponse).assertComplete()
    }

    @Test
    fun fetchAnimals_statusCodeNotOK_returnsPetfinderExceptionForStatusCode() {
        val request = setupFilterRequest()
        every { animalListResponse.animalList } returns listOf(mockk())
        val status = mockk<Status> {
            every { code } returns StatusCode.PFAPI_ERR_INTERNAL
        }
        every { header.status } returns status

        val observer = subject.fetchAnimals(request).test()

        observer.assertError(PetfinderException(StatusCode.PFAPI_ERR_INTERNAL))
    }

    @Test
    fun fetchAnimals_returnAnimalListNull_returnExceptionForNoAnimals() {
        val request = setupFilterRequest()
        val status = mockk<Status> {
            every { code } returns StatusCode.PFAPI_OK
        }
        every { header.status } returns status
        every { animalListResponse.animalList } returns null

        val observer = subject.fetchAnimals(request).test()

        observer.assertError(PetfinderException(StatusCode.ERR_NO_ANIMALS))
    }

    @Test
    fun fetchAnimals_returnAnimalListEmpty_returnExceptionForNoAnimals() {
        val request = setupFilterRequest()
        val status   = mockk<Status> {
            every { code } returns StatusCode.PFAPI_OK
        }
        every { header.status } returns status
        every { animalListResponse.animalList } returns emptyList()

        val observer = subject.fetchAnimals(request).test()

        observer.assertError(PetfinderException(StatusCode.ERR_NO_ANIMALS))
    }

    @Test
    fun fetchAnimal_animalIdFound_returnFetchedAnimalAndHeader() {
        val localAnimal = mockk<LocalAnimal> { every { id } returns 10 }
        val internetAnimal = mockk<InternetAnimal> { every { id } returns 10 }
        every { singleAnimalResponse.animal } returns internetAnimal
        every { animalService.fetchAnimal(any(), "10") } returns Single.just(singleAnimalResponse)

        val observer = subject.fetchAnimal(localAnimal).test()

        observer.assertValues(AnimalResponseWrapper(internetAnimal, header))
                .assertComplete()
        verify { observableHelper.applySingleSchedulers<Any>() }
    }

    @Test
    fun fetchAnimal_animalInResponseIsNull_returnOriginalAnimalAndHeader() {
        val localAnimal = mockk<LocalAnimal> { every { id } returns 10 }
        every { singleAnimalResponse.animal } returns null
        every { animalService.fetchAnimal(any(), "10") } returns Single.just(singleAnimalResponse)

        val observer = subject.fetchAnimal(localAnimal).test()

        observer.assertValues(AnimalResponseWrapper(localAnimal, header))
                .assertComplete()
        verify { observableHelper.applySingleSchedulers<Any>() }
    }

    private fun setupFilterRequest(): FetchAnimalsRequest {
        val filter = Filter().apply {
            sex = Sex.MALE
            age = Age.ADULT
            size = AnimalSize.LARGE
            breed = "Calico"
        }
        return FetchAnimalsRequest(AnimalType.CAT, 0, "zipcode", filter)
    }
}